sharktank/sharktank/utils/iree.py

```diff
@@ -397,3 +404,114 @@ def __call__(self, key: str, buffer_views: List[iree.runtime.HalBufferView]):
 
     return Callback(device)
+
+
+def trace_with_tracy(
+    fn: Callable[[int], Any],
+    /,
+    *,
+    output_trace_path: str = None,
+    port: int = None,
+    capture_extra_args: list[str] | None = None,
+) -> Any:
+    """Trace a callable with iree-tracy-capture.
+    The capture process is started before executing the tracing target.and is waited on
+    to finish. The traced target function is started in parallel.
+    If a port is not provided a free one is selected automatically.
+    """
+    capture_cmd = ["iree-tracy-capture", "-f"]
+    if output_trace_path:
+        capture_cmd += ["-o", output_trace_path]
+    if port is None:
+        from .io import find_free_port
+
+        port = find_free_port()
+    if capture_extra_args:
+        capture_cmd += capture_extra_args
+    capture_cmd += ["-p", f"{port}"]
+    with subprocess.Popen(capture_cmd) as capture_proc:
+        try:
+            res = fn(port)
+        except:
+            capture_proc.terminate()
+            raise
+        capture_process_return_code = capture_proc.wait()
+        if capture_process_return_code != 0:
+            raise subprocess.CalledProcessError(
+                f"Tracy capture process {capture_cmd} failed with return code {capture_process_return_code}"
+            )
+        return res
+
+
+def trace_command_with_tracy(
+    cmd: list[str],
+    /,
+    *,
+    output_trace_path: str = None,
+    port: int = None,
+    capture_extra_args: list[str] | None = None,
+    env: dict[str, str] | None = None,
+    **run_kwargs,
+):
+    """Trace an executable with Tracy."""
+
+    def fn(port: int):
+        env2 = env or os.environ
+        env2 = deepcopy(env2)
+        env2["TRACY_PORT"] = str(port)
+        proc = subprocess.run(cmd, env=env2, **run_kwargs)
+        proc.check_returncode()
+
+    trace_with_tracy(
+        fn,
+        output_trace_path=output_trace_path,
+        port=port,
+        capture_extra_args=capture_extra_args,
+    )
+
+
+def trace_model_with_tracy(
+    config: "ModelConfig", function: str, output_trace_path: str = None, **kwargs
+):
+    """Trace an already exported and compiled model with Tracy."""
+    cmd = [
+        sys.executable,
+        "-m",
+        "sharktank.tools.trace_model_with_tracy",
+        f"--function={function}",
+    ]
+    if output_trace_path is None:
+        output_trace_path = f"{config.iree_module_path}.tracy"
+    trace_command_with_tracy(
+        cmd,
+        input=json.dumps(config.asdict_for_saving()).encode(),
+        output_trace_path=output_trace_path,
+        **kwargs,
+    )
+
+
+def run_model_with_iree_run_module(
+    config: "ModelConfig", function: str, **subprocess_run_kwargs
+):
+    """Run an already exported and compiled model with iree-run-module.
+    It is required that is exports its input arguments.
+    """
+    cmd = [
+        "iree-run-module",
+        f"--module={config.iree_module_path}",
+        f"--device={config.iree_hal_driver}",
```

Comment:
Multidevice support in a later PR?

Reply:
Yes, at some point. The config has a `tensor_parallelism` parameter, that can be utilized for that.
The meaning of the `parameters_path` would need to be expanded to be a list as well. Or maybe a directory that contains the IRPA files. We need this anyway if we are to split the weights due to for example constraints of our storage like Hugging Face.